File: java_src/src/main/java/com/google/crypto/tink/hybrid/internal/HpkePrimitiveFactory.java
// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
///////////////////////////////////////////////////////////////////////////////

package com.google.crypto.tink.hybrid.internal;

import com.google.crypto.tink.proto.HpkeParams;
import com.google.crypto.tink.subtle.EllipticCurves;
import java.security.GeneralSecurityException;
import java.util.Arrays;

/**
 * Helper class for creating HPKE primitives from either algorithm identifiers or {@link
 * com.google.crypto.tink.proto.HpkeParams}.
 */
final class HpkePrimitiveFactory {
  /** Returns an {@link HpkeKem} primitive corresponding to {@code kemId}. */
  static HpkeKem createKem(byte[] kemId) throws GeneralSecurityException {
    if (Arrays.equals(kemId, HpkeUtil.X25519_HKDF_SHA256_KEM_ID)) {
      return new X25519HpkeKem(new HkdfHpkeKdf("HmacSha256"));
    } else if (Arrays.equals(kemId, HpkeUtil.P256_HKDF_SHA256_KEM_ID)) {
      return NistCurvesHpkeKem.fromCurve(EllipticCurves.CurveType.NIST_P256);
    } else if (Arrays.equals(kemId, HpkeUtil.P384_HKDF_SHA384_KEM_ID)) {
      return NistCurvesHpkeKem.fromCurve(EllipticCurves.CurveType.NIST_P384);
    } else if (Arrays.equals(kemId, HpkeUtil.P521_HKDF_SHA512_KEM_ID)) {
      return NistCurvesHpkeKem.fromCurve(EllipticCurves.CurveType.NIST_P521);
    }
    throw new IllegalArgumentException("Unrecognized HPKE KEM identifier");
  }

  /**
   * Returns an {@link HpkeKem} primitive corresponding to {@link
   * com.google.crypto.tink.proto.HpkeParams#getKem()}.
   */
  static HpkeKem createKem(HpkeParams params) throws GeneralSecurityException {
    if (params.getKem() == com.google.crypto.tink.proto.HpkeKem.DHKEM_X25519_HKDF_SHA256) {
      return new X25519HpkeKem(new HkdfHpkeKdf("HmacSha256"));
    } else if (params.getKem() == com.google.crypto.tink.proto.HpkeKem.DHKEM_P256_HKDF_SHA256) {
      return NistCurvesHpkeKem.fromCurve(EllipticCurves.CurveType.NIST_P256);
    } else if (params.getKem() == com.google.crypto.tink.proto.HpkeKem.DHKEM_P384_HKDF_SHA384) {
      return NistCurvesHpkeKem.fromCurve(EllipticCurves.CurveType.NIST_P384);
    } else if (params.getKem() == com.google.crypto.tink.proto.HpkeKem.DHKEM_P521_HKDF_SHA512) {
      return NistCurvesHpkeKem.fromCurve(EllipticCurves.CurveType.NIST_P521);
    }
    throw new IllegalArgumentException("Unrecognized HPKE KEM identifier");
  }

  /** Returns an {@link HpkeKdf} primitive corresponding to {@code kdfId}. */
  static HpkeKdf createKdf(byte[] kdfId) {
    if (Arrays.equals(kdfId, HpkeUtil.HKDF_SHA256_KDF_ID)) {
      return new HkdfHpkeKdf("HmacSha256");
    } else if (Arrays.equals(kdfId, HpkeUtil.HKDF_SHA384_KDF_ID)) {
      return new HkdfHpkeKdf("HmacSha384");
    } else if (Arrays.equals(kdfId, HpkeUtil.HKDF_SHA512_KDF_ID)) {
      return new HkdfHpkeKdf("HmacSha512");
    }
    throw new IllegalArgumentException("Unrecognized HPKE KDF identifier");
  }

  /**
   * Returns an {@link HpkeKdf} primitive corresponding to {@link
   * com.google.crypto.tink.proto.HpkeParams#getKdf()}.
   */
  static HpkeKdf createKdf(HpkeParams params) {
    if (params.getKdf() == com.google.crypto.tink.proto.HpkeKdf.HKDF_SHA256) {
      return new HkdfHpkeKdf("HmacSha256");
    } else if (params.getKdf() == com.google.crypto.tink.proto.HpkeKdf.HKDF_SHA384) {
      return new HkdfHpkeKdf("HmacSha384");
    } else if (params.getKdf() == com.google.crypto.tink.proto.HpkeKdf.HKDF_SHA512) {
      return new HkdfHpkeKdf("HmacSha512");
    }
    throw new IllegalArgumentException("Unrecognized HPKE KDF identifier");
  }

  /** Returns an {@link HpkeAead} primitive corresponding to {@code aeadId}. */
  static HpkeAead createAead(byte[] aeadId) throws GeneralSecurityException {
    if (Arrays.equals(aeadId, HpkeUtil.AES_128_GCM_AEAD_ID)) {
      return new AesGcmHpkeAead(16);
    } else if (Arrays.equals(aeadId, HpkeUtil.AES_256_GCM_AEAD_ID)) {
      return new AesGcmHpkeAead(32);
    } else if (Arrays.equals(aeadId, HpkeUtil.CHACHA20_POLY1305_AEAD_ID)) {
      return new ChaCha20Poly1305HpkeAead();
    }
    throw new IllegalArgumentException("Unrecognized HPKE AEAD identifier");
  }

  /**
   * Returns an {@link HpkeAead} primitive corresponding to {@link
   * com.google.crypto.tink.proto.HpkeParams#getAead()}.
   */
  static HpkeAead createAead(HpkeParams params) throws GeneralSecurityException {
    if (params.getAead() == com.google.crypto.tink.proto.HpkeAead.AES_128_GCM) {
      return new AesGcmHpkeAead(16);
    } else if (params.getAead() == com.google.crypto.tink.proto.HpkeAead.AES_256_GCM) {
      return new AesGcmHpkeAead(32);
    } else if (params.getAead() == com.google.crypto.tink.proto.HpkeAead.CHACHA20_POLY1305) {
      return new ChaCha20Poly1305HpkeAead();
    }
    throw new IllegalArgumentException("Unrecognized HPKE AEAD identifier");
  }

  private HpkePrimitiveFactory() {}
}

File: javascript/hybrid/internal/hpke/hpke_primitive_factory.ts
/**
 * @license
 * Copyright 2023 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import {InvalidArgumentsException} from '../../../exception/invalid_arguments_exception';
import {PbHpkeAead, PbHpkeKdf, PbHpkeKem, PbHpkeParams} from '../../../internal/proto';
import * as ellipticCurves from '../../../subtle/elliptic_curves';

import {AesGcmHpkeAead} from './aes_gcm_hpke_aead';
import {HkdfHpkeKdf} from './hkdf_hpke_kdf';
import {HpkeAead} from './hpke_aead';
import {HpkeKdf} from './hpke_kdf';
import {HpkeKem} from './hpke_kem';
import * as hpkeUtil from './hpke_util';
import {NistCurvesHpkeKem} from './nist_curves_hpke_kem';


/**
 * Helper class for creating HPKE primitives from either algorithm identifiers
 * or HpkeParams.
 */
export class HpkePrimitiveFactory {
  private constructor() {}

  /** Returns an HpkeKem primitive corresponding to the kemId. */
  static createKemFromId(kemId: Uint8Array): HpkeKem {
    switch (kemId) {
      case hpkeUtil.P256_HKDF_SHA256_KEM_ID:
        return NistCurvesHpkeKem.fromCurve(ellipticCurves.CurveType.P256);
      case hpkeUtil.P521_HKDF_SHA512_KEM_ID:
        return NistCurvesHpkeKem.fromCurve(ellipticCurves.CurveType.P521);
      default:
        throw new InvalidArgumentsException('Unrecognized HPKE KEM identifier');
    }
  }

  /** Returns an HpkeKem primitive corresponding to the given parameters. */
  static createKemFromParams(params: PbHpkeParams): HpkeKem {
    switch (params.getKem()) {
      case PbHpkeKem.DHKEM_P256_HKDF_SHA256:
        return NistCurvesHpkeKem.fromCurve(ellipticCurves.CurveType.P256);
      case PbHpkeKem.DHKEM_P521_HKDF_SHA512:
        return NistCurvesHpkeKem.fromCurve(ellipticCurves.CurveType.P521);
      default:
        throw new InvalidArgumentsException('Unrecognized HPKE KEM identifier');
    }
  }

  /** Returns an HpkeKdf primitive corresponding to the kdfId. */
  static createKdfFromId(kdfId: Uint8Array): HpkeKdf {
    switch (kdfId) {
      case hpkeUtil.HKDF_SHA256_KDF_ID:
        return new HkdfHpkeKdf('SHA-256');
      case hpkeUtil.HKDF_SHA512_KDF_ID:
        return new HkdfHpkeKdf('SHA-512');
      default:
        throw new InvalidArgumentsException('Unrecognized HPKE KDF identifier');
    }
  }

  /** Returns an HpkeKdf primitive corresponding to the given parameters. */
  static createKdfFromParams(params: PbHpkeParams): HpkeKdf {
    switch (params.getKdf()) {
      case PbHpkeKdf.HKDF_SHA256:
        return new HkdfHpkeKdf('SHA-256');
      case PbHpkeKdf.HKDF_SHA512:
        return new HkdfHpkeKdf('SHA-512');
      default:
        throw new InvalidArgumentsException('Unrecognized HPKE KDF identifier');
    }
  }

  /** Returns an HpkeAead primitive corresponding to the aeadId. */
  static createAeadFromId(aeadId: Uint8Array): HpkeAead {
    switch (aeadId) {
      case hpkeUtil.AES_128_GCM_AEAD_ID:
        return new AesGcmHpkeAead(16);
      case hpkeUtil.AES_256_GCM_AEAD_ID:
        return new AesGcmHpkeAead(32);
      default:
        throw new InvalidArgumentsException(
            'Unrecognized HPKE AEAD identifier');
    }
  }

  /** Returns an HpkeAead primitive corresponding to the given parameters. */
  static createAeadFromParams(params: PbHpkeParams): HpkeAead {
    switch (params.getAead()) {
      case PbHpkeAead.AES_128_GCM:
        return new AesGcmHpkeAead(16);
      case PbHpkeAead.AES_256_GCM:
        return new AesGcmHpkeAead(32);
      default:
        throw new InvalidArgumentsException(
            'Unrecognized HPKE AEAD identifier');
    }
  }
}
